Ensure custom content-types work

snazy opened the pull request:
* Add some more valdiations when registering content-type-bundles.
* Add a page on the project web site for registered content-types.
* Add tests in `:nessie-model` + `:nessie-versioned-spi` to verify that
  custom content-type-bundles and custom content-serializer-bundles
  work.

Fixes #6591

dimas-b commented on api/model/src/test/java/org/projectnessie/model/types/CustomTestContentTypeBundle.java:
```diff
@@ -0,0 +1,26 @@
+
+  @Override
+  public void register(Registrar registrar) {
+    registrar.register("TEST_CUSTOM_CONTENT_TYPE", CustomTestContent.class);
```
nit: use name constant from `CustomTestContent`?

dimas-b commented on api/model/src/test/java/org/projectnessie/model/types/CustomTestContentTypeBundle.java:
```diff
@@ -0,0 +1,26 @@
+
+import org.projectnessie.model.types.ContentTypes.Registrar;
+
+public class CustomTestContentTypeBundle implements ContentTypeBundle {
```
Suggestion: Do we really need the `ContentTypeBundle` interface? Why not load each custom type class via the service loader and then register them all automatically based on `@JsonTypeName` (which we validate to match the registered type name anyway)?

IIRC, service loader resources can list more than one service class per file.

With current approach it looks like type names leak across class boundaries without any good code-level reason :thinking: 

If that works, we might as well follow the same approach for `ContentSerializerBundle`, I guess.

snazy replied:
In theory, yes.
But we need `Class<? extends Content>` for the registration. Service loaders however return instantiated instances of the class.

snazy commented:
Updated the PR a bit.

* Introduced "top level" registry interfaces for content-type + content-serializer registration.
* Removed the `name` parameter

dimas-b commented on versioned/spi/src/main/java/org/projectnessie/versioned/store/ContentSerializerBundle.java:
```diff
@@ -15,9 +15,12 @@
+ */
 public interface ContentSerializerBundle {
-  void register(Consumer<ContentSerializer<?>> registry);
+  void register(ContentSerializerRegistry registry);
```
Well, unlike `Content.Type`, here we register serializer instances... Could you simply do `ServiceLoader.load(ContentSerializer.class)`? On the receiving side, we do not know specific type parameters anyway.

snazy replied:
But why have another approach here?

dimas-b replied:
Do remove an extra interface... but this is a minor point, optional.

dimas-b commented on api/model/src/main/java/org/projectnessie/model/types/ContentTypeRegistry.java:
```diff
@@ -0,0 +1,23 @@
+
+/** An implementation of this interface is passed to {@link ContentTypeBundle}s. */
+public interface ContentTypeRegistry {
+  void register(Class<? extends Content> type);
```
The Json part is fine, I think. My concern is only about the Open API spec (which has its own discriminators in the schema annotation on `Content`).

Since we now expect custom types to be present, is it worth declaring a sub-set of possible types as `oneOf` in our Open API spec?

snazy replied:
Define "possible types" ;)

The problem is that we cannot dynamically extend that `oneOf` + the OpenAPI discriminators. OTOH, it's only "an issue" for new/completely-custom-private types. New "core" types will naturally be added there.

dimas-b replied:
Why would we want to keep `oneOf` in the Open API spec at all? With pluggable types, the spec is guaranteed to be skewed sooner or later. 

I'm proposing to replace it with a generic `Object` type (or the Open API) equivalent.

adutra replied:
I think this could become a problem for the future Event API. The generated JSON schema for this API currently has this:

```json
      "Content" : {
        "title" : "Content",
        "type" : "object",
        "properties" : {
          "id" : {
            "type" : "string"
          }
        },
        "oneOf" : [ {
          "$ref" : "#/components/schemas/IcebergTable"
        }, {
          "$ref" : "#/components/schemas/DeltaLakeTable"
        }, {
          "$ref" : "#/components/schemas/IcebergView"
        }, {
          "$ref" : "#/components/schemas/Namespace"
        } ],
        "discriminator" : {
          "propertyName" : "type",
          "mapping" : {
            "ICEBERG_TABLE" : "#/components/schemas/IcebergTable",
            "DELTA_LAKE_TABLE" : "#/components/schemas/DeltaLakeTable",
            "ICEBERG_VIEW" : "#/components/schemas/IcebergView",
            "NAMESPACE" : "#/components/schemas/Namespace"
          }
        }
      },
```

The `oneOf` approach will be problematic if someone uses this schema to attempt a polymorphic deserialization of components of different types, for example when using `KafkaJsonSchemaDeserializer` to deserialize events from a Kafka topic.

Even if the consumer has the jar containing the custom content implementation, the schema won't allow it, and deserializing such a content is likely to fail imho.

(However, I _think_ this could work when using a schema-less deserializer like `KafkaJsonDeserializer`.)

snazy replied:
Hm - custom content-types aren't new though. Not sure why events infra needs to know the whole model.

snazy replied:
> perhaps replace with anyOf(coreType1, coreType2, ..., Object).

`anyOf` seems to work

dimas-b commented on api/model/src/main/java/org/projectnessie/model/Content.java:
```diff
@@ -34,7 +35,13 @@
+      DeltaLakeTable.class,
+      IcebergView.class,
+      Namespace.class,
+      Map.class
```
nit: this does not have a name in Swagger / Open API yaml:
```
    Content_V2:
      title: Content
      type: object
      properties:
        id:
          type: string
      anyOf:
      - $ref: '#/components/schemas/IcebergTable_V2'
      - $ref: '#/components/schemas/DeltaLakeTable_V2'
      - $ref: '#/components/schemas/IcebergView_V2'
      - $ref: '#/components/schemas/Namespace_V2'
      - type: object
```
not a big deal, but perhaps we can name it "Custom Content" (latger)

snazy replied:
Can't put a "name" for it :(
